openssl.go

```diff
@@ -0,0 +1,185 @@
+// +build openssl
+
+package libp2ptls
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	ic "github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-openssl"
+)
+
+// libp2pNID is the NID for the libp2p-tls oid.
+var libp2pNID openssl.NID = openssl.CreateObjectIdentifier(getExtensionString(extensionID),
+	"libp2p-tls",
+	"Object Identifier for libp2p-tls")
+
+// openSSLIdentity is used to
+type openSSLIdentity struct {
+	certificate *openssl.Certificate
+	privateKey  openssl.PrivateKey
+}
+
+// newOpenSSLIdentity create openSSLIdentity.
+func newOpenSSLIdentity(sk ic.PrivKey) (*openSSLIdentity, error) {
+	cert, privKey, err := createOpenSSLCertificate(sk)
+	if err != nil {
+		return nil, err
+	}
+	return &openSSLIdentity{
+		certificate: cert,
+		privateKey:  privKey,
+	}, nil
+}
+
+func newIdentity(sk ic.PrivKey) (identity, error) {
+	return newOpenSSLIdentity(sk)
+}
+
+// CreateServerConn creates server connection to do the tls handshake.
+func (o *openSSLIdentity) CreateServerConn(insecure net.Conn) (handshakeConn,
+	<-chan ic.PubKey, error) {
+	opensslCtx, keyCh, err := o.createOpenSSLCtx("")
+	if err := opensslCtx.UsePrivateKey(o.privateKey); err != nil {
+		return nil, nil, err
+	}
+	conn, err := openssl.Server(insecure, opensslCtx)
+	if err != nil {
+		return nil, nil, err
+	}
+	return conn, keyCh, nil
+}
+
+// CreateClientConn creates client connection to do the tls handshake.
+func (o *openSSLIdentity) CreateClientConn(insecure net.Conn, remote peer.ID) (handshakeConn,
+	<-chan ic.PubKey, error) {
+	opensslCtx, keyCh, err := o.createOpenSSLCtx(remote)
+
+	conn, err := openssl.Client(insecure, opensslCtx)
+	if err != nil {
+		return nil, nil, err
+	}
+	return conn, keyCh, nil
+}
+
+func (o *openSSLIdentity) createOpenSSLCtx(remote peer.ID) (*openssl.Ctx,
+	<-chan ic.PubKey, error) {
+	keyCh := make(chan ic.PubKey, 4)
+	opensslCtx, err := openssl.NewCtxWithVersion(openssl.TLSv1_2)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Add the certificate.
+	if err := opensslCtx.UseCertificate(o.certificate); err != nil {
+		return nil, nil, err
+	}
+
+	// Enable two way tls.
+	opensslCtx.SetVerifyMode(openssl.VerifyPeer)
```

Comment:
I'm not familiar with the go-openssl API here. Why `VerifyPeer` and not `VerifyFailIfNoPeerCert`? What happens if the client doesn't present a certificate at all? And what is `VerifyClientOnce` anyway?

Reply:
See the [openssl man page]( https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_verify.html#NOTES). The go-openssl verify option bindings have a one-to-one correspondence to the C openssl bindings. I set it to both `VerifyPeer` and `VerifyFailIfNoPeerCert`.